qutebrowser/completion/strategies/strategy.py

```diff
@@ -0,0 +1,48 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2014-2020 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Classes that return miscellaneous completion models."""
+
+from typing import Sequence, Optional, Any, Callable, Tuple
+from abc import ABC, abstractmethod
+
+from qutebrowser.completion.completionmodel import CompletionModel
+from qutebrowser.completion.completer import CompletionInfo
+
+
+class DeletionUnsupportedError(Exception):
+    """Raises when the completion strategy does not support deletion."""
+
+
+class CompletionStrategy(ABC):
+    COLUMN_WIDTHS: Tuple[int, int, int] = (20, 70, 10)
+
+    def __init__(self):
+        self.model = CompletionModel(column_widths=self.COLUMN_WIDTHS)
+
+    def __call__(self, *args: Any, **kwds: Any) -> Optional[CompletionModel]:
+        return self.populate(*args, **kwds)
+
+    @abstractmethod
+    def populate(self, *args: str, info: Optional[CompletionInfo]) -> Optional[CompletionModel]:
```

Comment:
I think this should always return a `CompletionModel`, even if it is empty. That avoids the mypy problem you have above:

```
Item "None" of "Optional[CompletionModel]" has no attribute "set_pattern"
```

Reply:
I also don't think `info` is `Optional`

Reply:
If we're using `__call__`, let's make `populate` private (`_populate`) and just invoke that from `__call__`.

Reply:
One reason I hesitated to do that was because using the ```__call__``` syntax in ```test_models.py``` would have been ugly and hard to read. What do you think?

Reply:
I think it looks fine, but I don't feel strongly. We shouldn't have two ways to call it though. Lets choose either populate or call, not both.

Reply:
Based on the discussion above (about not creating `self.model` in `__init__`), I think we do need two methods - I've called it `populate` in my initial proposal, but with the new distinction, that's a bit confusing. I'd thus propose to switch to:

- `_populate` (the thing implemented by subclasses)
- `complete` (the thing called by the completer to use the model) - this one could be `__call__` instead. Other names could be `run`, `generate`, `get_model`, `to_model`, but I personally like `complete` best. No strong feelings either way, though (but `run` and `generate` seem a bit generic).

I originally liked the idea of `__call__` because I thought there was one clear action for those completion strategies (since they evolved from functions), but now I think this isn't true anymore. I also agree that `model = urlmodel.URL()(info=info)` is weird - it's a bit better if splitting it up into two lines, but still, `model = urlmodel.URL().complete(info=info)` makes it clear what's going on.